*** gcc/gnattools/configure.orig	2011-02-13 14:29:15.000000000 +0000
--- gcc/gnattools/configure	2011-02-12 19:47:33.000000000 +0000
***************
*** 2111,2116 ****
--- 2111,2121 ----
      mlib-tgt-specific.adb<mlib-tgt-specific-lynxos.adb \
      indepsw.adb<indepsw-gnu.adb"
      ;;
+     arm*-*-eabi)
+     TOOLS_TARGET_PAIRS="\
+     mlib-tgt-specific.adb<mlib-tgt-specific-xi.adb \
+     indepsw.adb<indepsw-gnu.adb"
+     ;;
  esac
  
  # From user or toplevel makefile.
