*** gcc-4.6-20110129.orig/gcc/ada/gcc-interface/Makefile.in	2010-10-25 16:28:13.000000000 +0100
--- gcc-4.6-20110129/gcc/ada/gcc-interface/Makefile.in	2011-02-03 18:46:51.000000000 +0000
***************
*** 1577,1583 ****
    LIBRARY_VERSION := $(subst .,_,$(LIB_VERSION))
  endif
  
! ifeq ($(strip $(filter-out avr none powerpc% eabispe leon% erc32% unknown elf,$(targ))),)
    TOOLS_TARGET_PAIRS=\
    mlib-tgt-specific.adb<mlib-tgt-specific-xi.adb \
    indepsw.adb<indepsw-gnu.adb
--- 1577,1583 ----
    LIBRARY_VERSION := $(subst .,_,$(LIB_VERSION))
  endif
  
! ifeq ($(strip $(filter-out avr arm none powerpc% eabi eabispe leon% erc32% unknown elf,$(targ))),)
    TOOLS_TARGET_PAIRS=\
    mlib-tgt-specific.adb<mlib-tgt-specific-xi.adb \
    indepsw.adb<indepsw-gnu.adb
