*** gcc.orig/Makefile.in.orig	2011-02-06 19:00:53.000000000 +0000
--- gcc/Makefile.in	2011-02-06 19:02:10.000000000 +0000
***************
*** 60527,60533 ****
  all-stageprofile-libcpp: maybe-all-stageprofile-intl
  all-stagefeedback-libcpp: maybe-all-stagefeedback-intl
  all-fixincludes: maybe-all-libiberty
! all-gnattools: maybe-all-target-libada
  all-lto-plugin: maybe-all-libiberty
  
  all-stage1-lto-plugin: maybe-all-stage1-libiberty
--- 60527,60533 ----
  all-stageprofile-libcpp: maybe-all-stageprofile-intl
  all-stagefeedback-libcpp: maybe-all-stagefeedback-intl
  all-fixincludes: maybe-all-libiberty
! all-gnattools:
  all-lto-plugin: maybe-all-libiberty
  
  all-stage1-lto-plugin: maybe-all-stage1-libiberty
